### PR TITLE
Move zip to build.d

### PR DESCRIPTION
**Pull request by marler8997**

### CODE REVIEW

File: src/build.d [hunk 1]
```diff
@@ -513,6 +514,25 @@ alias checkwhitespace = memoize!(function()
     return new DependencyRef(dep);
 });
 
+alias zip = memoize!(function()
+{
+    Dependency dep;
+    with (dep)
+    {
+        name = "zip";
+        target = "dmdsrc.zip";
```
Comment (MoonlightSentinel) [inline src/build.d]: Absolute path?

File: src/build.d [hunk 2]
```diff
@@ -513,6 +514,25 @@ alias checkwhitespace = memoize!(function()
...
+        sources = [thisBuildScript].chain(.sources.root, .sources.backend, .sources.lexer,
```
Comment (MoonlightSentinel) [inline src/build.d]: `thisBuildScript` is already an implicit dependency, see
https://github.com/dlang/dmd/blob/442ba964ce07a183ec5d4a6bc13daba083c88bd1/src/build.d#L1346

File: src/build.d [hunk 3]
```diff
@@ -513,6 +514,25 @@ alias checkwhitespace = memoize!(function()
...
+            .sources.frontendHeaders, .sources.dmd).array;
+        msg = "ZIP " ~ target;
+        commandFunction = () {
+            if (exists(target))
+                remove(target);
+            run([env["ZIP"], target.stripExtension] ~ sources);
```
Comment (MoonlightSentinel) [inline src/build.d]: Why strip the extension?

**Review by MoonlightSentinel — COMMENTED** [review]
Some small nits/questions

File: src/build.d (hunk 3 above)
Comment (marler8997) [reply src/build.d]: just copying what the makefiles are doing

Comment (marler8997) [reply src/build.d]: I've removed the `stripExtension`, well see if it passes the tests

File: src/build.d (hunk 2 above)
Comment (marler8997) [reply src/build.d]: Right, but we still need it in the zip command, I"ll just add it in the command instead then.

File: src/build.d (hunk 1 above)
Comment (marler8997) [reply src/build.d]: Yeah, based on the Makefile I was thinking it was generated the zip file in the "working directory", but that's not the case.

File: src/win32.mak [hunk 4]
```diff
@@ -334,13 +332,8 @@ detab:
...
 	$(TOLF) $(SRCS) $(GLUESRC) $(ROOTSRC) $(BACKSRC) $(MAKEFILES)
 
-zip: detab tolf $(MAKEFILES)
-	$(DEL) dmdsrc.zip
-	$(ZIP) dmdsrc $(MAKEFILES)
-	$(ZIP) dmdsrc $(SRCS)
-	$(ZIP) dmdsrc $(GLUESRC)
-	$(ZIP) dmdsrc $(BACKSRC)
-	$(ZIP) dmdsrc $(ROOTSRC)
+zip: $(GEN)\build.exe
```
Comment (MoonlightSentinel) [inline src/win32.mak]: Missing dependencies:
```suggestion
zip: detab tolf $(GEN)\build.exe
```

File: src/build.d [hunk 5]
```diff
@@ -513,6 +514,25 @@ alias checkwhitespace = memoize!(function()
...
+    {
+        name = "zip";
+        target = srcDir.buildPath("dmdsrc.zip");
+        sources = .sources.root.chain(.sources.backend, .sources.lexer,
+            .sources.frontendHeaders, .sources.dmd).array;
+        msg = "ZIP " ~ target;
+        commandFunction = () {
+            if (exists(target))
+                remove(target);
+            run([env["ZIP"], target] ~ sources ~ [thisBuildScript]);
```
Comment (MoonlightSentinel) [inline src/build.d]:
```suggestion
            run([env["ZIP"], target, thisBuildScript] ~ sources);
```

File: src/build.d [hunk 6]
```diff
@@ -513,6 +514,25 @@ alias checkwhitespace = memoize!(function()
...
 });
 
+alias zip = memoize!(function()
+{
+    Dependency dep;
+    with (dep)
+    {
+        name = "zip";
+        target = srcDir.buildPath("dmdsrc.zip");
+        sources = .sources.root.chain(.sources.backend, .sources.lexer,
```
Comment (MoonlightSentinel) [inline src/build.d]: Simple array concatenation would involve less noise (but that is propably a matter of personal preference)

**Review by MoonlightSentinel — APPROVED** [review]
Looks good otherwise

**Comment by marler8997** [comment]
@MoonlightSentinel good suggestions, all integrated

**Review by PetarKirov — APPROVED** [review]
Nice!

**Pull request merged**